### PR TITLE
PXT-1047 <fn_name>.to_sql implementation for common string operations

### PR DESCRIPTION
**Pull request by christopherpestano**
Adds to_sql implementations for 9 string UDFs, enabling PostgreSQL pushdown instead of Python-side evaluation:

- Regex: contains_re, count, match, fullmatch, replace_re — using PG ~, ~*, regexp_count, regexp_replace. Falls back to Python when flags or n parameters are used.
- Character classification: isascii — using octet_length/char_length comparison. Other is* functions (isalnum, isalpha, isdecimal, isdigit, islower, isspace, isupper) were intentionally excluded because PG POSIX classes are ASCII-only under the C locale, causing silent mismatches with Python's Unicode-aware behavior.
- Padding: ljust, rjust — using lpad/rpad.
- Numeric: zfill — sign-aware zero-padding via lpad.

Tests cover SQL vs Python equivalence for all new implementations with edge cases including empty strings, emoji, CJK characters, accented text, regex metacharacters, alternation anchoring, case-insensitive mode, and Python fallback paths.

### CODE REVIEW

**Comment by aaron-siegel** [comment]
Hmmm, I'm not sure about this. When we first discussed this idea, I'd envisioned adding `to_sql` methods for Python functions that have direct (or nearly direct) SQL equivalents. I hadn't envisioned writing a bunch of bespoke SQL logic to replicate Python functionality that doesn't have a SQL analog.

The risk is that the bespoke logic misses certain edge cases (or handles them subtly differently from Python), leading to inconsistent application behavior. This risk is heightened if we're relying heavily on AI to generate the custom logic. I'd say we should discuss the scope of this effort as a team before you invest too much more in it.

**Comment by christopherpestano** [comment]
Point taken, talked to Sergey and Marcel in office about this, I think the path I want to take this down now is to focus purely on string operations such as regex and those that are easily representable as regex. This is a pretty common use case and though it will require some bespoke logic, it shouldn't be too crazy. Narrowing the scope here will help focus on what will likely be the highest impact to actual use cases, and avoid bad user experience with common string ops.

File: pixeltable/functions/string.py [hunk 1]
```diff
@@ -106,6 +106,13 @@ def contains_re(self: str, pattern: str, flags: int = 0) -> bool:
     return bool(re.search(pattern, self, flags))
 
 
+@contains_re.to_sql
+def _(self: sql.ColumnElement, pattern: sql.ColumnElement, flags: sql.ColumnElement | None = None) -> sql.ColumnElement:
+    if flags is not None:
+        return None  # Can't handle arbitrary re flags in SQL
+    return self.op('~')(pattern)
```
[the remark this answers is not in the document]
Comment (christopherpestano) [reply pixeltable/functions/string.py:218]: This pattern is consistent with the rest of the codebase. The pre-existing replace.to_sql and slice.to_sql use the same is not None check for their optional parameters (lines 630, 780). When a user omits the argument, Pixeltable passes None to the to_sql function, not the Python default value. Explicitly passing flags=0 is unlikely since it is the default, and introducing BindParameter introspection adds complexity without meaningful benefit.

File: pixeltable/functions/string.py [hunk 2]
```diff
@@ -118,6 +125,13 @@ def count(self: str, pattern: str, flags: int = 0) -> int:
     return builtins.len(re.findall(pattern, self, flags))
 
 
+@count.to_sql
+def _(self: sql.ColumnElement, pattern: sql.ColumnElement, flags: sql.ColumnElement | None = None) -> sql.ColumnElement:
+    if flags is not None:
+        return None  # Can't handle arbitrary re flags in SQL
+    return sql.func.regexp_count(self, pattern)
```
[the remark this answers is not in the document]
Comment (christopherpestano) [reply pixeltable/functions/string.py:239]: Same rationale as above. This follows the existing codebase pattern.

File: pixeltable/functions/string.py [hunk 3]
```diff
@@ -223,6 +237,23 @@ def fullmatch(self: str, pattern: str, case: bool = True, flags: int = 0) -> boo
...
 
+@fullmatch.to_sql
+def _(
+    self: sql.ColumnElement,
+    pattern: sql.ColumnElement,
+    case: sql.ColumnElement | None = None,
+    flags: sql.ColumnElement | None = None,
+) -> sql.ColumnElement:
+    if flags is not None:
+        return None  # Can't handle arbitrary re flags in SQL
```
[the remark this answers is not in the document]
Comment (christopherpestano) [reply pixeltable/functions/string.py]: Same rationale as above. This follows the existing codebase pattern.

File: pixeltable/functions/string.py
[the remark this answers is not in the document]
Comment (christopherpestano) [reply pixeltable/functions/string.py:602]: Same rationale as above. This follows the existing codebase pattern.

[the remark this answers is not in the document]
Comment (christopherpestano) [reply pixeltable/functions/string.py:752]: Same rationale as above. This follows the existing codebase pattern.

File: pixeltable/functions/string.py [hunk 4]
```diff
@@ -281,6 +319,12 @@ def isdecimal(self: str) -> bool:
     return self.isdecimal()
 
 
+@isdecimal.to_sql
+def _(self: sql.ColumnElement) -> sql.ColumnElement:
+    # At least one character and all are decimal digits (0-9)
+    return sql.and_(sql.func.char_length(self) > 0, self.op('~')('^[0-9]+$'))
+
+
```
[the remark this answers is not in the document]
Comment (christopherpestano) [reply pixeltable/functions/string.py]: Removed the SQL pushdown. The embedded PostgreSQL server uses LC_CTYPE='C' which only recognizes ASCII digits, but Python str.isdecimal() is fully Unicode-aware (Arabic-Indic, Devanagari, fullwidth digits, etc.). No way to detect at query planning time whether the data contains non-ASCII characters, so the pushdown would silently return wrong results.

File: pixeltable/functions/string.py [hunk 5]
```diff
@@ -354,6 +398,12 @@ def isspace(self: str) -> bool:
     return self.isspace()
 
 
+@isspace.to_sql
+def _(self: sql.ColumnElement) -> sql.ColumnElement:
+    # At least one character and all characters are whitespace
+    return sql.and_(sql.func.char_length(self) > 0, self.op('~')('^[[:space:]]+$'))
+
+
```
[the remark this answers is not in the document]
Comment (christopherpestano) [reply pixeltable/functions/string.py]: Same fix as isdecimal. Removed the SQL pushdown since LC_CTYPE='C' only recognizes ASCII whitespace, but Python str.isspace() handles NBSP, en space, ideographic space, etc.

File: tests/functions/test_string.py [hunk 6]
```diff
@@ -328,6 +332,305 @@ def test_format(self, uses_db: None) -> None:
...
+            '\u4e16\u754c123',
+            '\U0001f600\U0001f600\U0001f600',
+        ]
+        validate_update_status(t.insert({'s': s} for s in test_strs), expected_rows=len(test_strs))
+
+        test_params = [(isascii, str.isascii), (isdecimal, str.isdecimal), (isspace, str.isspace)]
+
+        for pxt_fn, str_fn in test_params:
+            # SQL execution
+            res_sql = t.select(out=pxt_fn(t.s)).collect()['out']
```
[the remark this answers is not in the document]
Comment (christopherpestano) [reply tests/functions/test_string.py]: Added Unicode divergence test cases: Arabic-Indic digit, Devanagari digit, fullwidth digit, NBSP, en space, and ideographic space. These now pass correctly since isdecimal and isspace run through Python (SQL pushdowns removed).

**Comment by christopherpestano** [comment]
Note: the isdecimal and isspace SQL pushdowns were removed because the embedded PostgreSQL server uses LC_CTYPE='C', which limits POSIX character classes to ASCII. We should look into switching the embedded pgserver to LC_CTYPE='en_US.UTF-8' (or similar) so these pushdowns can be re-enabled with full Unicode support.

File: tests/functions/test_string.py [hunk 7]
```diff
@@ -328,6 +332,311 @@ def test_format(self, uses_db: None) -> None:
...
+            assert res_sql == expected, f'zfill w={w} SQL: {res_sql} != {expected}'
+            assert res_py == expected, f'zfill w={w} Py: {res_py} != {expected}'
+
+    def test_is_validation_functions(self, uses_db: None) -> None:
+        """Test is* validation functions with edge cases including Unicode.
+
+        isdecimal and isspace have no SQL pushdown because the embedded PostgreSQL server uses
+        LC_CTYPE='C', which only recognizes ASCII characters for POSIX character classes.
+        Python's str.isdecimal() and str.isspace() are fully Unicode-aware, so SQL pushdown
+        would silently return wrong results for non-ASCII inputs.
```
[the remark this answers is not in the document]
Comment (christopherpestano) [reply tests/functions/test_string.py]: Fixed.

File: tests/functions/test_string.py [hunk 8]
```diff
@@ -328,6 +332,311 @@ def test_format(self, uses_db: None) -> None:
...
+
+        def check_sql_and_py(pxt_fn: pxt.Function, *args: Any, **kwargs: Any) -> tuple[list[Any], list[Any]]:
+            """Return (sql_results, python_results) for pxt_fn applied to the test table."""
+            res_sql = t.select(out=pxt_fn(t.s, *args, **kwargs)).collect()['out']
+            res_py = t.select(out=pxt_fn(t.s.apply(lambda x: x, col_type=pxt.String), *args, **kwargs)).collect()['out']
+            return res_sql, res_py
+
+        # ── contains_re ───────────────────────────────────────────────────────
+        # Basic literal pattern
+        for pat in ['cat', 'dog', 'hello']:
```
Comment (sergey-mkhitaryan) [inline tests/functions/test_string.py]: some patterns that do not get translated correctly, and cause this test to fail: `r'\bcat\b'`, `r'cat\B'`, `(?P<cat>cat)`, `r'\0'` (to repro this one, add `\x00` to `test_strs`)

**Review by sergey-mkhitaryan — CHANGES_REQUESTED** [review]
